Fixed missing escaping/security related issues Codeinwp/ppom-pro#195

girishpanchal30 opened the pull request:
Issue Codeinwp/ppom-pro#195

mskapusuz commented on classes/fields.class.php:
```diff
@@ -211,7 +211,7 @@ function render_field_settings() {
 		}
 
 		$html .= '</div>';
-		echo $html;
+		echo ppom_esc_html( $html );
 	}
```
we should not rely on built-in `ppom_esc_html`

if some elements are not necessarily; we can optimize/fix `ppom_esc_html` or have a new context for `wp_kses_allowed_html`

thats valid all codebase which `ppom_esc_html` is used.

**an example output:**
![Screen Shot 2023-04-26 at 14 22 35](https://user-images.githubusercontent.com/25361626/234560431-85807b43-ef0b-44c2-b41a-7d66302079da.jpg)

mskapusuz replied:
I would like to confirm whether it is necessary to allow scripts and other elements. Could you please check and share your findings with me? Once we have the results, we can move forward.

girishpanchal30 replied:
@ mskapusuz I think we are not using script and style tags. Should I remove it from the `ppom_esc_html` function?
Ref: https://github.com/Codeinwp/woocommerce-product-addon/blob/master/inc/validation.php#L46

mskapusuz replied:
Today in later hours I'll continue to my reviews

girishpanchal30 replied:
@mskapusuz Removed script/style from the `ppom_esc_html` function

mskapusuz commented on inc/admin.php:
```diff
@@ -61,9 +61,9 @@ function ppom_admin_product_meta_column( $column, $post_id ) {
 				);
-				echo sprintf( __( '<a href="%1$s">%2$s</a>', 'woocommerce-product-addon' ), $url_edit, $ppom->meta_title );
+				echo sprintf( esc_html__( '<a href="%1$s">%2$s</a>', 'woocommerce-product-addon' ), esc_url( $url_edit ), esc_html( $ppom->meta_title ) );
 			} else {
-				echo sprintf( __( '<a class="btn button" href="%1$s">%2$s</a>', 'woocommerce-product-addon' ), $ppom_settings_url, 'Add Fields' );
+				echo sprintf( esc_html__( '<a class="btn button" href="%1$s">%2$s</a>', 'woocommerce-product-addon' ), esc_url( $ppom_settings_url ), esc_html_e( 'Add Fields', 'woocommerce-product-addon' ) );
```
`esc_html` breaks anchor element

mskapusuz commented on inc/admin.php:
```diff
@@ -47,10 +47,10 @@ function ppom_admin_product_meta_column( $column, $post_id ) {
 						);
-						echo sprintf( __( '<a href="%1$s">%2$s</a>', 'woocommerce-product-addon' ), $url_edit, $meta_title );
+						echo sprintf( esc_html__( '<a href="%1$s">%2$s</a>', 'woocommerce-product-addon' ), esc_url( $url_edit ), esc_html( $meta_title ) );
 						echo ', ';
 					} else {
-						echo sprintf( __( '<a class="btn button" href="%1$s">%2$s</a>', 'woocommerce-product-addon' ), $ppom_settings_url, 'Add Fields' );
+						echo sprintf( esc_html__( '<a class="btn button" href="%1$s">%2$s</a>', 'woocommerce-product-addon' ), esc_url( $ppom_settings_url ), esc_html_e( 'Add Fields', 'woocommerce-product-addon' ) );
```
`esc_html` breaks anchor element

mskapusuz commented on templates/admin/ppom-fields.php:
```diff
@@ -244,22 +242,22 @@ class="ppom-modal-box ppom-slider ppom_sort_id_<?php echo esc_attr( $f_index );
 								<div class="ppom-modal-body">
 									<?php
-									echo $form_meta->render_field_meta( $defualt_fields, $field_type, $f_index, $field_meta );
+										echo $form_meta->render_field_meta( $defualt_fields, $field_type, $f_index, $field_meta ); // phpcs:ignore
```
the [render_field_meta](https://github.com/Codeinwp/woocommerce-product-addon/blob/bugfix/pro/195/classes/fields.class.php#L300) method output does not have escaping.

I think we can make a refactor here.

Let's move the `echo` to inside of the `render_field_meta` do not return any value and remove the ignore announcement. Please make sure you've refactor all of the places which calls `render_field_meta`

girishpanchal30 replied:
If we move echo to the inside of the render_field_meta it might issue here https://github.com/Codeinwp/woocommerce-product-addon/blob/bcbb51af430404e3c7ac8f49a081508287fd94ef/classes/fields.class.php#L202

And I've add phpcs:ignore becuase of `wp_kses_post` function automatically remove `class` from input tag
Ref: https://tinyurl.com/2afyk8j9 - It breaks the structure

Please let me know if you have any idea about this issue

mskapusuz replied:
how about if we have echo in `render_field_settings` too?

mskapusuz reviewed (changes requested):
@girishpanchal30 I've left some comments, I'll another look in later hours when I'm available. Thank you!

mskapusuz commented:
Hey @girishpanchal30 

I think you've implemented the changes and these are ready for the review again, right?

Thanks

mskapusuz commented on templates/admin/ppom-fields.php:
```diff
@@ -26,8 +26,9 @@
 $ppom_field_index       = 1;
 
+// phpcs:ignore WordPress.Security.NonceVerification.Recommended
 if ( isset( $_REQUEST ['productmeta_id'] ) && $_REQUEST ['do_meta'] == 'edit' ) {
-
+	// phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
that's valid for other nonce ignore announcements too if there are

mskapusuz commented on templates/frontend/component/quantities/grid-layout.php:
```diff
@@ -101,15 +101,15 @@ class="ppom-quantity"
 						data-usebase_price="<?php echo esc_attr( $usebaseprice ); ?>"
 						data-unlink_qty="<?php echo esc_attr( $unlink_qty ); ?>"
-						<?php echo apply_filters( 'ppom_fe_form_element_custom_attr', '', $fm ); ?>
+						<?php echo apply_filters( 'ppom_fe_form_element_custom_attr', '', $fm ); // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped ?>
 						data-price="<?php echo esc_attr( $the_price ); ?>"
```
we should not have ignore announcement for the apply_filters call. We cannot rely on the data which comes by the hook.

So, we should surround it with a proper escaping function.

That's valid for all `ppom_fe_form_element_custom_attr` apply_filters call in the codebase. (There are near to 23 results)

mskapusuz commented on templates/frontend/inputs/checkbox.php:
```diff
@@ -140,11 +130,11 @@ class="<?php echo esc_attr( $input_class ); ?>"
 						data-data_name="<?php echo esc_attr( $fm->data_name() ); ?>"
 						value="<?php echo esc_attr( $key ); ?>"
-						<?php echo $checked_option; ?>
+						<?php echo $checked_option // phpcs:ignore; ?>
 				>
```
that's valid for all ignore announcements for $checked_option variables in other inputs.

mskapusuz commented on templates/frontend/inputs/select.php:
```diff
@@ -41,21 +41,20 @@
 	<!-- if title of field exist -->
 	<?php if ( $fm->field_label() ) : ?>
-		<label class="<?php echo esc_attr( $fm->label_classes() ); ?>"
-			   for="<?php echo esc_attr( $fm->data_name() ); ?>"><?php echo $fm->field_label(); ?></label>
+		<label class="<?php echo esc_attr( $fm->label_classes() ); ?>" for="<?php echo esc_attr( $fm->data_name() ); ?>"><?php echo wp_kses_post( $fm->field_label() ); ?></label>
```
for the field label, do we need `wp_kses_post`? can we use `esc_html` instead?

mskapusuz commented on templates/input/quantities.php:
```diff
@@ -160,7 +160,7 @@
 					   data-max="<?php echo esc_attr( $args['max_qty'] ); ?>"
 					   data-label="<?php echo esc_attr( $label ); ?>"
 					   data-includeprice="<?php echo esc_attr( $include_productprice ); ?>"
-					   name="<?php echo htmlentities( $name ); ?>" type="number" class="ppom-quantity"
+					   name="<?php echo esc_html( htmlentities( $name ) ); ?>" type="number" class="ppom-quantity"
 					   data-usebase_price="<?php echo esc_attr( $usebaseprice ); ?>"
```
I think we can use `esc_attr` here

mskapusuz reviewed (changes requested):
@girishpanchal30 thanks for the fixes, generally looks good 🚀 I've left some comments, please let me know if does it makes sense to you.

Thank you!

girishpanchal30 commented:
Hey @mskapusuz,

Thank you for review the code, fixed/resolved all changes.
Please review it again when you have a moment.

Thanks

mskapusuz commented on templates/frontend/inputs/measure.php:
```diff
@@ -84,7 +83,7 @@ class="<?php echo esc_attr( $input_classes ); ?>"
 				<?php
 				// Add input extra attributes
 				foreach ( $input_attr as $key => $val ) {
-					echo $key . '="' . $val . '"';
+					echo sprintf( '%s="%s"', esc_html( $key ), esc_html( $val ) );
 				}
```
I think, using `esc_attr` instead of `esc_html` would be better here (valid other inputs too)

mskapusuz commented on templates/frontend/inputs/palettes.php:
```diff
@@ -51,16 +51,15 @@
     }';
 
 echo '<style>';
-echo $custom_css;
+echo $custom_css; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 echo '</style>';
```
can we remove the ignore announcement?

mskapusuz commented on templates/frontend/inputs/select.php:
```diff
@@ -101,7 +100,7 @@ class="<?php echo esc_attr( $option_class ); ?>"
 					data-without_tax="<?php echo esc_attr( $without_tax ); ?>"
 					data-data_name="<?php echo esc_attr( $fm->data_name() ); ?>"
 					data-option_weight="<?php echo esc_attr( $option_weight ); ?>"
-					<?php echo $selected_value; ?>
+					<?php echo esc_attr( $selected_value ); ?>
```
`esc_attr` breaks selected prop output here

![Screen Shot 2023-04-28 at 10 22 57](https://user-images.githubusercontent.com/25361626/235081965-fca5caf0-fe59-4e49-a4b5-d158ca7c2e7e.jpg)

mskapusuz commented on templates/input/quantities.php:
```diff
@@ -32,15 +32,15 @@
 							<?php
 							$the_price = isset( $opt['price'] ) && $opt['price'] != '' ? $opt['price'] : $default_price;
 							if ( $the_price ) {
-								echo ' <span class="ppom-quantity-price-wrap">' . wc_price( $the_price ) . '</span>';
+								echo ' <span class="ppom-quantity-price-wrap">' . esc_html( wc_price( $the_price ) ) . '</span>';
 							}
```
I think that was not resolved. That suggestion does makes sense to you @girishpanchal30 Please let me know if I'm missing something.

mskapusuz replied:
that's valid for other `esc_html( wc_price() )` usages in the PR

mskapusuz replied:
sample render:

![Screen Shot 2023-04-28 at 10 29 37](https://user-images.githubusercontent.com/25361626/235083356-c311dfef-ede7-4189-bfe7-af5e96389116.jpg)
![Screen Shot 2023-04-28 at 10 29 32](https://user-images.githubusercontent.com/25361626/235083358-70be94e4-b5e9-4460-bf05-9caa2c8c2c23.jpg)
![Screen Shot 2023-04-28 at 10 28 44](https://user-images.githubusercontent.com/25361626/235083360-4f6552e7-c68f-4d79-9d74-a96ac658d0fa.jpg)
![Screen Shot 2023-04-28 at 10 28 37](https://user-images.githubusercontent.com/25361626/235083363-ce399cf0-fd19-4cc4-b929-7aff92602284.jpg)

mskapusuz reviewed (changes requested):
Hey @girishpanchal30 

Thanks for resolving the suggestions. 🚀 I've left some things as far I can see. Apart from that, I suggest please test the PR on visually before the QA, as I noticed some part of the UI broken. E.g: on the front side, the labels are broken. Another e.g: visit WP admin -> visit PPOM group -> add a field -> save, you'll see the group is shown as empty. (no fields)

![Screen Shot 2023-04-28 at 10 46 32](https://user-images.githubusercontent.com/25361626/235086998-6c0c33b3-8e13-4171-9409-406131aeeef3.jpg)

mskapusuz commented on classes/admin.class.php:
```diff
@@ -369,7 +369,7 @@ class="woocommerce-help-tip" data-tip="<?php echo $value['desc']; ?>"></span></l
 					<?php
 					if ( ! empty( $selected_roles ) ) {
 						foreach ( $selected_roles as $key => $val ) {
-							echo '<option value="' . esc_attr( $key ) . '"' . wc_selected( $key, $selections ) . '>' . esc_html( $val ) . '</option>'; // WPCS: XSS ok.
+							echo '<option value="' . esc_attr( $key ) . '"' . wc_selected( $key, $selections ) . '>' . esc_html( $val ) . '</option>'; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 						}
```
let's remove `phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped`

girishpanchal30 commented:
@mskapusuz Thank you for your feedback, fixed other mentioned issues/suggestions

>  I noticed some part of the UI broken. E.g: on the front side, the labels are broken. 

Due to the removed style/script tag support, it remove some inline style.
Ref: https://github.com/Codeinwp/woocommerce-product-addon/pull/139/commits/2daa5f0d513d7b75c8af2cfe21fd2dc0339901b7

Now I've added style tag support to fix the broken layout issues.

Thanks

mskapusuz reviewed (approved):
LGTM 🚀 Thank you!

irinelenache commented:
@girishpanchal30 Did a short test and here is what i found: 

- [ ] The Audio/video inputs are not working anymore https://vertis.d.pr/v/rTfeCo
- [ ] The Bulk quantity input is visually broken and the max value is ignored
before this PR: https://vertis.d.pr/i/zPPwst
after the PR: https://vertis.d.pr/i/pKbjHT

- [ ] There is an `s` character visible in the Checkout input https://vertis.d.pr/i/h0IuHf
- [ ] The phone input has some code that's printed in frontend https://vertis.d.pr/i/MSzvrV
- [ ] ^ Same for the color palette https://vertis.d.pr/i/kOjxQI
- [ ] ^ Same for the color picker https://vertis.d.pr/i/TOimrU

Can you please use this instance and check if things in frontend look the same after updating with the PR? There were a lot of easy to spot regressions and i stopped testing after a few PPOM fields 🙏 
```
https://demolition-raco.vertisite.cloud/wp-admin
demo-ppom-lite
9qh0qRtX
```

girishpanchal30 commented:
Hey @irinelenache,

Thank you for the review, I've fixed all mentioned issue.

Please check with free and pro plugin PRs
Free: https://github.com/Codeinwp/woocommerce-product-addon/pull/139

>  The phone input has some code that's printed in frontend https://vertis.d.pr/i/MSzvrV

Pro: https://github.com/Codeinwp/ppom-pro/pull/205

Thanks

irinelenache commented:
@girishpanchal30 Please check if **all the fields** in the provided instance look as they should in frontend and then send it to testing. After another test i found the image field is broken https://vertis.d.pr/v/Tsvt4A

HardeepAsrani commented:
@girishpanchal30 We had a discussion and we can put this on hold for now as it might result in a lot of regressions. Thanks for the work so far.
